src/shogun/kernel/SpectralMixtureKernel.cpp

```diff
@@ -0,0 +1,221 @@
+{
+	init();
+	CDotKernel::init(l, r);
+	init_components(num_components);
+}
+
+CSpectralMixtureKernel::CSpectralMixtureKernel(const SGVector<float64_t>& weights,
+	const SGVector<float64_t>& means, const SGVector<float64_t>& stddeviations,
+	int32_t size) : CDotKernel(size)
+{
+	init();
+	set_component_parameters(weights, means, stddeviations);
+}
+
+void CSpectralMixtureKernel::init()
+{
+	SG_ADD(&m_weights, "weights",
+		"Vector of weights for each mixture component", MS_AVAILABLE, GRADIENT_AVAILABLE);
+	SG_ADD(&m_means, "means",
+		"Vector of means for each mixture component", MS_AVAILABLE, GRADIENT_AVAILABLE);
+	SG_ADD(&m_stddeviations, "stddeviations",
+		"Vector of standard deviations for each mixture component", MS_AVAILABLE, GRADIENT_AVAILABLE);
+}
+
+bool CSpectralMixtureKernel::init_components(int32_t num_components)
+{
+	REQUIRE(has_features(), "Features not set - cannot initialize components.");
+
+	m_weights=SGVector<float64_t>(num_components);
+	m_weights.set_const(1.0/num_components);
+
+	m_means=SGVector<float64_t>(num_components);
+	m_means.set_const(0.0);
+
+	m_stddeviations=SGVector<float64_t>(num_components);
+	m_stddeviations.set_const(1.0);
+
+	return true;
+}
+
+int32_t CSpectralMixtureKernel::get_num_components() const
+{
+	return m_weights.size();
+}
+
+void CSpectralMixtureKernel::set_component_parameters(const SGVector<float64_t>& weights,
+	const SGVector<float64_t>& means, const SGVector<float64_t>& stddeviations)
+{
+	bool same_size= (weights.size()==means.size()) && (weights.size()==stddeviations.size());
+	REQUIRE(same_size, "Weights, means, and stddeviations vectors must be of equal size.");
+
+	m_weights = weights;
+	m_means = means;
+	m_stddeviations = stddeviations;
+}
+
+float64_t CSpectralMixtureKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	REQUIRE(has_features(), "Features not set - cannot compute kernel value.");
+	REQUIRE(m_weights.size() > 0, "Mixture components not set - cannot compute kernel values.");
+
+	CDotFeatures* dotlhs = dynamic_cast<CDotFeatures*>(lhs);
+	CDotFeatures* dotrhs = dynamic_cast<CDotFeatures*>(rhs);
+
+	REQUIRE(dotlhs != NULL, "Left-hand-side features must be of type CDotFeatures");
+	REQUIRE(dotrhs != NULL, "Right-hand-side features must be of type CDotFeatures");
+
+	float64_t result = 0.0;
+	float64_t distance_sq = dotlhs->dot(idx_a, dotlhs, idx_a) + dotrhs->dot(idx_b, dotrhs, idx_b) - 2*dotlhs->dot(idx_a, dotrhs, idx_b);
+
+	for(int i=0; i<m_weights.size(); i++)
+		result += CMath::pow(m_weights[i],2)*CMath::exp(-2*CMath::pow(CMath::PI,2)*distance_sq*CMath::pow(m_stddeviations[i],2))*CMath::cos(2*CMath::PI*CMath::sqrt(distance_sq)*m_means[i]);
+
+	return result;
+}
+
+SGVector<float64_t> CSpectralMixtureKernel::get_squared_features(CDotFeatures* features)
+{
+	SGVector<float64_t> sqr_features = SGVector<float64_t>(features->get_num_vectors());
+	
+	for (int i=0; i<sqr_features.size(); i++)
+		sqr_features[i] = features->dot(i, features, i);
```

Comment:
seems like this could be done via linalg more efficiently too (openmp, GPU, etc)
@lambday